osd: object class loading and execution permissions

noahdesu opened the pull request:
This adds options 'osd class load list' that is a list of classes allowed to be loaded (or * for all), and 'osd class default list' which is a list of classes (or * for all) that may be used without explicit permission.

liewegas commented on src/test/osd/osdcap.cc:
```diff
@@ -270,6 +270,11 @@ TEST(OSDCap, BasicRW) {
+  // class whitelist flag works
+  ASSERT_TRUE(cap.is_capable("bar", "", 0, "foo", false, true, false, true, true));
+  ASSERT_TRUE(cap.is_capable("bar", "", 0, "foo", true, false, true, false, true));
+  ASSERT_TRUE(cap.is_capable("bar", "", 0, "foo", true, true, true, true, true));
```
This doesn't look right... did you try runing this unit test?

The "allow rw" behavior shouldn't change.  But the "allow x" case should change.. where existing test cases where whitelisting is false should go from true -> false, and new cases should only be true if whitelist is true...

noahdesu replied:
My understanding was that if a class was whitelisted, then irregardless of having 'allow x' it should be permitted to call into an object class. So this test is supposed to show that with only 'allow rw' but a whitelisted class (last parameter: true) then may_class_read and may_class_write are OK. Have I got these semantics messed up?

liewegas replied:
Yeah, we want something a bit different:

1. if you have 'allow x' or 'class-read' or 'class-write', then you can 
execute any method on any *whitelisted* class.  We'll whitelist existing 
stuff by, so there is effectively no net change here (no new permissions 
and nothing is lost).

2. if you have 'allow class foo', then you can use class foo regardless of 
whether it is whitelisted or not (provided the osd is loading it). 

@athanatos right?

athanatos replied:
That seems to make sense.

noahdesu replied:
does 'allow *' override the whitelist?

liewegas replied:
yeah, 'allow *' is like root.

On Tue, 28 Jun 2016, Noah Watkins wrote:

> 
> In src/test/osd/osdcap.cc:
> 
> > @@ -270,6 +270,11 @@ TEST(OSDCap, BasicRW) {
> >    ASSERT_FALSE(cap.is_capable("bar", "", 0, "foo", false, true, false, true));
> >    ASSERT_FALSE(cap.is_capable("bar", "", 0, "foo", true, false, true, false));
> >    ASSERT_FALSE(cap.is_capable("bar", "", 0, "foo", true, true, true, true));
> > +
> > +  // class whitelist flag works
> > +  ASSERT_TRUE(cap.is_capable("bar", "", 0, "foo", false, true, false, true, true));
> > +  ASSERT_TRUE(cap.is_capable("bar", "", 0, "foo", true, false, true, false, true));
> > +  ASSERT_TRUE(cap.is_capable("bar", "", 0, "foo", true, true, true, true, true));
> 
> does 'allow *' override the whitelist?
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AAabh_n9-cs0YNVIW8IwkQQRxb2aCuOoks5qQW4agaJpZM4I_hOr.gif]
> 
> 
>

liewegas commented:
The change looks right except for the unit test.  I think we'll want a whole slew of cases for everywhere either of op_may_class_{read,write} is true with the whitelist false and true cases, where a false changes the is_capable result from true -> false.

noahdesu commented:
I've pushed up v2 for review. This has 'allow *' allowed to call any object class method, enforces whitelist, and supports the 'allow class foo' format. I updated the tests per your suggestion, and added new tests for the 'allow class foo' cases.

noahdesu commented:
It just occurred to me that this patch doesn't handle loading class libraries that are not listed in 'osd class load list' unless it is set to `*`. Should we use 'allow class foo' to permit loading of non-default classes?

liewegas commented:
It seems okay to me to require the admin to either load * or list it 
explicitly.

noahdesu commented:
Cool. That makes sense.

liewegas commented:
This looks good to me!  @athanatos want to take a final look?

athanatos commented on src/osd/PG.cc:
```diff
@@ -1870,7 +1870,9 @@ bool PG::op_has_sufficient_caps(OpRequestRef& op)
-			     op->need_class_write_cap());
+			     op->need_class_write_cap(),
+			     op->class_name(),
+			     op->class_whitelisted());
```
An op could invoke two different classes, right?

noahdesu replied:
Oh I had only considered the case of classes calling other classes in which case they were transitively permitted. Are you referring to using multiple ObjectOperation::exec?

athanatos replied:
The librados ObjectOperation api lets you perform multiple execs in the same Operation.

athanatos commented:
A librados op can have more than one class operation, right?  The OpRequestRef machinery should reflect that.

noahdesu commented:
@athanatos ok thanks for catching that problem. i pushed a new version for review that handles multiple execs in the same operation (i didn't squash... seems easier for reviews to not?).

athanatos commented:
Ok, I think that looks right!

noahdesu commented:
@tchaikov i was just getting ready to squash the top 3 commits, but it looks like `wip-kefu-testing` might have this branch close to merging? Would it be easier for you to sqash them?

tchaikov commented:
@noahdesu as long as the squashed code is identical to the one before squashing, they are the same to me. =)

noahdesu commented:
@tchaikov i pushed up the new version with the top 3 commits squashed. code is identical to before.

tchaikov commented:
tested at 
- http://pulpito.ceph.com/kchai-2016-07-12_11:14:46-rados-wip-kefu-testing2---basic-mira/
- http://pulpito.ceph.com/kchai-2016-07-12_11:14:46-rados-wip-kefu-testing2---basic-mira/

the failed tests are addressed by #10306.

tchaikov commented on src/osd/ClassHandler.h:
```diff
@@ -101,9 +101,12 @@ class ClassHandler
+  ClassData *_get_class(const string& cname, bool check_allowed);
   int _load_class(ClassData *cls);
 
+  bool in_class_list(const std::string& cname,
```
nit, could be a static member function.

tchaikov commented on src/osd/ClassHandler.cc:
```diff
@@ -80,14 +83,44 @@ void ClassHandler::shutdown()
+  auto it = std::find_first_of(begin, end,
+      targets.begin(), targets.end());
+
+  if (it == end)
```
could just `return it != end`.

tchaikov commented:
@noahdesu besides some nits, i think it might be worthy to have this features in the PendingReleaseNotes. but i don't want to block you, i am merging it, since sage and sam have given your their blessings.

noahdesu commented:
@tchaikov i will submit separate PRs for pending release notes and nits. They aren't blockers. Thanks.
